### DISCUSSION TITLE
Using #let to define aliases for commonly used math expressions

### DISCUSSION THREAD

**Discussion by aldanor** [typst/typst#3357] [General]
Maybe I'm missing something completely obvious, but I did spend time going through docs and some examples in the source code as well (which didn't help much).

I'm trying to do something seemingly simple -- define some commonly used 'math expression' (commonly used piece of content in `$$` context), e.g. something like `a0 := alpha_0`. Out of everything I've tried, the two approaches (see below) seemed to have sort-of-worked, but I'm not sure that's the correct way of doing it.

It also does result in a weird "boxing" of the resulting expressions, which messes up subscripts and superscripts. Here's an example:

```typst
#let a0v1 = sym.alpha + sub([0])
#let a0v2 = $alpha_0$
$ &alpha_0^2\ &a0v1^2\ &a0v2^2 $
```

which results in

![image](https://github.com/typst/typst/assets/2418513/e43a6d5e-658f-4481-bb5a-ad5be8c1f00c)

Is there a way to define something like `alpha_0` via `#let` so it works correctly (i.e. as it shows on the first line in the example above)? (and if what I'm doing is completely wrong, maybe the docs should reflect it in a more obvious fashion?)

Thanks!

**Comment by davidleejy** [comment]
Was defining a few of my own math-related snippets with the let(.) function, but did not realize this 🥲.

**Comment by Julian-Wassmann** [comment]
```typst
#let a0v1 = sym.alpha + sub([0])
```
When defining it this way, you're not using an equation. It is regular content with subscript, which is the reason it looks different from `$alpha_0^2$`.

```typst
#let a0v2 = $alpha_0$
$ a0v2^2 $
```
This way, the variable is an actual equation, which is the right way of doing it. However, in the equation output, the variable is wrapped with another function (^2). This is why the ^2 appears behind your defined variable.

The best way is to put the base and the exponent in the same equation:
```typst
#let a02 = $alpha_0^2$
```
This will then output with the correct math style.

But maybe this is not flexible enough, because you have different exponents and don't want to define a new variable for each variant. In this case, you can use a function which outputs an equation:
```typst
#let a0 = exp => $alpha_0^exp$
$ a0(2) + a0(3) $
```

**Comment by aldanor** [comment]
> But maybe this is not flexible enough, because you have different exponents and don't want to define a new variable for each variant. In this case, you can use a function which outputs an equation

Thanks - this is probably the closest solution. As a newcomer to Typst, I must admit the whole thing is a little bit counterintuitive (especially the 'boxing' part -- it sort of makes sense but is not obvious).

I guess my question is then -- is there no way to 'inject' superscripts/subscripts etc into an expression defined via let other than defining a function? Kind of like "inline function" that gets expanded before it gets rendered? (not really a 'macro', more like a lazy expansion)

**Reply by Julian-Wassmann** [reply]
I haven't found one. I tried using [math.attach](https://typst.app/docs/reference/math/attach#functions-attach), but that isn't working. Maybe someone else can help here?

**Comment by Enivex** [comment]
There is currently no great way to do this. You can technically do a show rule on `math.attach`, but it's very hard to get it to work correctly in all situations.

It is definitely a thing that is desirable though, and it's very likely that it'll change at some point in the future.

**Comment by davidleejy** [comment]
Came across https://castel.dev/post/lecture-notes-1/#sub--and-superscripts and was wondering if popular addons like ultisnips and hypersnips that provide customizable snippets might come in handy in this case.

A snippet for expanding `a0` into `alpha_0`, or more generally, `[a]\d` (regex) into `alpha_\d` appears possible.

Disclaimer: I've not used snippets extensively but am considering a foray into them.

**Reply by aldanor** [reply]
@davidleejy It's definitely possible to use your editor's "expand-snippet" functionality (e.g. one of the snips packages for vim), but that's not the question here. If you have a long and repetitive expression, you might want your typst source file to look cleaner and more concise, hence the original question of trying to do it via `#let`.

**Reply by davidleejy** [reply]
Thanks for the suggestion.

> but that's not the question here

Ah, I was under the initial impression that you might be trying to save keystrokes by getting the equivalent of `alpha_0` with `a0`. The article I linked was written by a math PhD student who devised snippets to convert expressions of the form `a0` into $a_0$ in order to take math notes at speed during lectures.

> want your typst source file to look cleaner and more concise, hence the original question of trying to do it via #let.

If shortening the .typ file is primary reason for your question rather than speeding up typing, perhaps you might like to consider keyboards that possess multiple layers, ideally with each layer housing symbols that appear in your document?

I'm currently using [Andrew Myer's keyboard configuration](https://www.cs.cornell.edu/andru/release/lattice-kbd/) for math and logic related documents.

The following is a sample of my typst code and output. Observe that the greek and mathematical symbols used are represented as single characters in the typst code (e.g., α, Λ, ≠, ⋃, ∅) instead of strings like `alpha`, `Lambda`, `union.big`, `emptyset`.

<img width="437" alt="Screenshot 2024-02-10 at 3 21 18 AM" src="https://github.com/typst/typst/assets/1690072/da9b7a86-c05a-4c01-91fd-39b11564ffa2">

```typst
#definition("Compact")[

Let $X$ be a topological space.

Let $X$ be #underline[compact], then T.F.A.E.:

+ #underline[Every] #underline[open cover] of $X$ admits a #underline[finite subcover]. That is, for every open cover ${U_α}_(α ∈ Λ)$ of $X$, there exists a #underline[finite] indexing subset $Λ_0 ⊆ Λ$ such that $⋃_(α ∈ Λ_0) U_α #text(red)[=] X.$ \
  #[
  #set par(hanging-indent: 1em)
  Note:\
    By "admits a finite subcover" it means that the finite open subcover is a subset of the open cover ${U_α}_(α ∈ Λ_0) ⊆ {U_α}_(α ∈ Λ)$.
  ]
  

+ Every collection $cC$ of #underline[closed] subsets of $X$ with the #underline[finite intersection property] (F.I.P.) has a #underline[nonempty] intersection $⋂_(C ∈ cC) C ≠ ∅.$\
  #[
  #set par(hanging-indent: 1em)
  Formally:\
    $∀ "collection" cC = {U_α}_(α ∈ Λ) "of closed subsets of" X$, $∀ "finite indexing set" Λ_0 ⊆ Λ$, \ 
    $⋂_(α ∈ Λ_0) C_α ≠ ∅$ ~~~$==>$~~~ $⋂_(α ∈ Λ) C_α ≠ ∅.$
  ]
  This property is the same as Proposition@prop-compactness-in-terms-of-FIP which defines compactness in terms of the F.I.P.

]
```

**Comment by davidleejy** [comment]
@aldanor 

While waiting for a solution using `#let` or another Typst construct, I'd like to suggest considering text expansion "expand-snippet" methods again.

> It's definitely possible to use your editor's "expand-snippet" functionality (e.g. one of the snips packages for vim), but that's not the question here. If you have a long and repetitive expression, you might want your typst source file to look cleaner and more concise, hence the original question of trying to do it via #let.

While I had previously suggested a text expansion of "a0" to "alpha_0", it is also possible to expand ";a;" to "α" (the trigger, here ";a;", is up to the user to define) – achieving (i) a cleaner & more concise typst source file, and (ii) identical printing compared to "alpha_0". This screen capture shows text expansion in action and the output:

https://github.com/typst/typst/assets/1690072/6dfad185-2c9c-41e2-865a-71d56ed01758

In the text expansion software Espanso (https://github.com/espanso/espanso), one can configure it so that when the characters ";a;" are typed, the entire string is immediately replaced with a single greek character "α".
